### ISSUE TITLE
Hashset add returns true instead of false

### ISSUE THREAD

**Issue by Nazgaul** [nhibernate/nhibernate-core#2476]
Hi.
Kind hard to describe the scenario but ill do my best.
this is nhibernate v5.2.7 screenshot
![Annotation 2020-08-12 172636](https://user-images.githubusercontent.com/3366179/90028338-3ccc2780-dcc2-11ea-9b77-5adff6d3a5c6.png)
this is nhibernate v5.3.2 screenshot
![Annotation 2020-08-12 172852](https://user-images.githubusercontent.com/3366179/90028374-46ee2600-dcc2-11ea-9e83-037615c8ca59.png)

now _followers is a hashset of followers
if i do a foreach before i do the add it return false as expected.
On 5.3.2 i don't see any select to bring the collection...

the mapping in user is
```c#
  HasMany(x => x.Followers).Access.CamelCaseField(Prefix.Underscore)
                .Cascade.AllDeleteOrphan()
                .KeyColumn("UserId").Inverse().AsSet();
```
and
```c#
       HasMany(x => x.Following).KeyColumn("FollowerId")
                .Cascade.AllDeleteOrphan().AsSet();
```
again this is hard scenario due to complex domain module. 
I can give you any more info that you'd like.

**Comment by maca88** [comment]
Can you provide the id mapping of Follow class? Does `Follow` class override `Equals` and `GetHashCode` methods?

**Comment by Nazgaul** [comment]
` public FollowMap()
        {
            Id(x => x.Id).GeneratedBy.GuidComb();
            References(x => x.User).Column("UserId").Not.Nullable().UniqueKey("c_follower");
            References(x => x.Follower).Column("FollowerId").Not.Nullable().UniqueKey("c_follower");
            Map(x => x.Created).Insert().Not.Update();
            Map(x => x.Subscriber).Nullable();
            Table("UsersRelationship");
        }`

Follow implementation of equal methods

`  public virtual bool Equals(Follow other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return Equals(User.Id, other.User.Id) && Equals(Follower.Id, other.Follower.Id);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Follow)obj);
        }

        public override int GetHashCode()
        {
            return (User.Id.GetHashCode() * 101) ^ (Follower.Id.GetHashCode() * 107);
        }`

I go from one entity ( course) and then go to course.tutor.user.addFollwer(user)
And if i do a foreach before ( to bring the collection ) the hashset is working ( and i see a select is generated).
on 5.3 - the select is not generated.

**Comment by maca88** [comment]
As I expected, the problem here is that `Follow` class uses `Id` property as the identifier but the overridden `Equals` and `GetHashCode` use `User` and `Follower` instead. This scenario will not work with `lazy="extra"` even with 5.2.x, because the element existence check performed on the database would check the `Id` property which in this case is zero. The reason why in version `5.2.x` works with `lazy="true"` and not in `5.3.x`, is because the optimization done in #2010 to avoid initializing the collection in case a transient element is added to the collection. I made a fix in #2481, which fallbacks to the previous behavior in case the element overrides `Equals` method.

**Comment by Nazgaul** [comment]
Thanks allot of identifying the problem

**Comment by fredericDelaporte** [comment]
Fixed with #2481
